File: hypersim-sdk/src/clients/WebSocketClient.ts
/**
 * WebSocket streaming client for real-time HyperEVM and HyperCore data
 */

import WebSocket from 'ws';
import { EventEmitter } from 'events';
import { Network } from '../types/network';
import { NetworkError, ConnectionError } from '../types/errors';
import { HYPERCORE_ENDPOINTS } from '../utils/constants';

/**
 * WebSocket subscription types
 */
export type SubscriptionType = 
  | 'trades'
  | 'book'
  | 'bbo'
  | 'userFills'
  | 'userOrders'
  | 'userEvents'
  | 'candle'
  | 'l2Book'
  | 'notification';

/**
 * WebSocket subscription request
 */
export interface WSSubscription {
  type: SubscriptionType;
  coin?: string;
  user?: string;
  interval?: string;
}

/**
 * WebSocket message structure
 */
export interface WSMessage {
  channel: string;
  data: any;
  timestamp?: number;
}

/**
 * WebSocket client configuration
 */
export interface WebSocketClientConfig {
  /** Network to connect to */
  network: Network;
  /** Custom WebSocket endpoint */
  wsEndpoint?: string;
  /** Connection timeout in milliseconds */
  connectionTimeout?: number;
  /** Reconnect interval in milliseconds */
  reconnectInterval?: number;
  /** Max reconnection attempts */
  maxReconnectAttempts?: number;
  /** Heartbeat interval in milliseconds */
  heartbeatInterval?: number;
  /** Enable debug logging */
  debug?: boolean;
  /** Enable compression */
  compression?: boolean;
}

/**
 * Connection states
 */
export enum ConnectionState {
  DISCONNECTED = 'disconnected',
  CONNECTING = 'connecting',
  CONNECTED = 'connected',
  RECONNECTING = 'reconnecting',
  ERROR = 'error'
}

/**
 * WebSocket streaming client for HyperEVM real-time data
 */
export class WebSocketClient extends EventEmitter {
  private readonly config: Required<WebSocketClientConfig>;
  private ws: WebSocket | null = null;
  private connectionState: ConnectionState = ConnectionState.DISCONNECTED;
  private subscriptions = new Set<string>();
  private reconnectAttempts = 0;
  private heartbeatTimer?: NodeJS.Timeout;
  private reconnectTimer?: NodeJS.Timeout;
  private connectionTimer?: NodeJS.Timeout;
  private lastHeartbeat = 0;
  private messageQueue: any[] = [];
  private metrics = {
    messagesReceived: 0,
    messagesSent: 0,
    reconnections: 0,
    errors: 0,
    lastActivity: 0
  };

  constructor(config: WebSocketClientConfig) {
    super();
    
    this.config = {
      network: config.network,
      wsEndpoint: config.wsEndpoint || this.getDefaultEndpoint(config.network),
      connectionTimeout: config.connectionTimeout || 10000,
      reconnectInterval: config.reconnectInterval || 5000,
      maxReconnectAttempts: config.maxReconnectAttempts || 10,
      heartbeatInterval: config.heartbeatInterval || 30000,
      debug: config.debug || false,
      compression: config.compression || true
    };

    if (this.config.debug) {
      console.log('[WebSocket Client] Initialized with config:', {
        network: this.config.network,
        endpoint: this.config.wsEndpoint,
        compression: this.config.compression
      });
    }
  }

  /**
   * Connect to WebSocket endpoint
   */
  async connect(): Promise<void> {
    if (this.connectionState === ConnectionState.CONNECTED) {
      if (this.config.debug) {
        console.log('[WebSocket Client] Already connected');
      }
      return;
    }

    if (this.connectionState === ConnectionState.CONNECTING) {
      if (this.config.debug) {
        console.log('[WebSocket Client] Connection already in progress');
      }
      return;
    }

    this.setConnectionState(ConnectionState.CONNECTING);

    try {
      await this.establishConnection();
    } catch (error) {
      this.setConnectionState(ConnectionState.ERROR);
      throw new ConnectionError(`Failed to connect: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    if (this.config.debug) {
      console.log('[WebSocket Client] Disconnecting...');
    }

    this.clearTimers();
    this.reconnectAttempts = 0;
    
    if (this.ws) {
      this.ws.removeAllListeners();
      this.ws.close(1000, 'Normal closure');
      this.ws = null;
    }

    this.setConnectionState(ConnectionState.DISCONNECTED);
    this.subscriptions.clear();
    this.messageQueue = [];
  }

  /**
   * Subscribe to a data stream
   */
  async subscribe(subscription: WSSubscription): Promise<void> {
    const subKey = this.getSubscriptionKey(subscription);
    
    if (this.subscriptions.has(subKey)) {
      if (this.config.debug) {
        console.log(`[WebSocket Client] Already subscribed to: ${subKey}`);
      }
      return;
    }

    const message = {
      method: 'subscribe',
      subscription
    };

    await this.sendMessage(message);
    this.subscriptions.add(subKey);

    if (this.config.debug) {
      console.log(`[WebSocket Client] Subscribed to: ${subKey}`);
    }
  }

  /**
   * Unsubscribe from a data stream
   */
  async unsubscribe(subscription: WSSubscription): Promise<void> {
    const subKey = this.getSubscriptionKey(subscription);
    
    if (!this.subscriptions.has(subKey)) {
      if (this.config.debug) {
        console.log(`[WebSocket Client] Not subscribed to: ${subKey}`);
      }
      return;
    }

    const message = {
      method: 'unsubscribe',
      subscription
    };

    await this.sendMessage(message);
    this.subscriptions.delete(subKey);

    if (this.config.debug) {
      console.log(`[WebSocket Client] Unsubscribed from: ${subKey}`);
    }
  }

  /**
   * Send a message to the WebSocket
   */
  private async sendMessage(message: any): Promise<void> {
    if (this.connectionState !== ConnectionState.CONNECTED) {
      this.messageQueue.push(message);
      if (this.config.debug) {
        console.log('[WebSocket Client] Message queued (not connected)');
      }
      return;
    }

    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      this.messageQueue.push(message);
      if (this.config.debug) {
        console.log('[WebSocket Client] Message queued (WebSocket not ready)');
      }
      return;
    }

    try {
      const serialized = JSON.stringify(message);
      this.ws.send(serialized);
      this.metrics.messagesSent++;
      
      if (this.config.debug) {
        console.log('[WebSocket Client] Message sent:', message);
      }
    } catch (error) {
      console.error('[WebSocket Client] Failed to send message:', error);
      this.messageQueue.push(message);
    }
  }

  /**
   * Process queued messages
   */
  private processMessageQueue(): void {
    if (this.messageQueue.length === 0) {
      return;
    }

    if (this.config.debug) {
      console.log(`[WebSocket Client] Processing ${this.messageQueue.length} queued messages`);
    }

    const messages = [...this.messageQueue];
    this.messageQueue = [];

    messages.forEach(message => {
      this.sendMessage(message).catch(error => {
        console.error('[WebSocket Client] Failed to send queued message:', error);
      });
    });
  }

  /**
   * Establish WebSocket connection
   */
  private async establishConnection(): Promise<void> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.config.wsEndpoint, {
        handshakeTimeout: this.config.connectionTimeout,
      });

      this.connectionTimer = setTimeout(() => {
        ws.terminate();
        reject(new Error('Connection timeout'));
      }, this.config.connectionTimeout);

      ws.on('open', () => {
        this.clearConnectionTimer();
        this.ws = ws;
        this.setConnectionState(ConnectionState.CONNECTED);
        this.reconnectAttempts = 0;
        this.setupHeartbeat();
        this.processMessageQueue();
        
        if (this.config.debug) {
          console.log('[WebSocket Client] Connected successfully');
        }
        
        resolve();
      });

      ws.on('message', (data) => {
        this.handleMessage(data);
      });

      ws.on('close', (code, reason) => {
        this.handleClose(code, reason.toString());
      });

      ws.on('error', (error) => {
        this.clearConnectionTimer();
        this.handleError(error);
        reject(error);
      });

      ws.on('ping', () => {
        if (this.config.debug) {
          console.log('[WebSocket Client] Received ping');
        }
        ws.pong();
      });
    });
  }

  /**
   * Handle incoming WebSocket message
   */
  private handleMessage(data: WebSocket.Data): void {
    try {
      const message = JSON.parse(data.toString());
      this.metrics.messagesReceived++;
      this.metrics.lastActivity = Date.now();
      
      if (message.type === 'heartbeat') {
        this.lastHeartbeat = Date.now();
        return;
      }

      if (this.config.debug) {
        console.log('[WebSocket Client] Message received:', message);
      }

      // Emit message event with proper structure
      const wsMessage: WSMessage = {
        channel: message.channel || 'unknown',
        data: message.data || message,
        timestamp: message.timestamp || Date.now()
      };

      this.emit('message', wsMessage);
      
      // Emit specific event types
      if (message.channel) {
        this.emit(message.channel, wsMessage.data);
      }
    } catch (error) {
      console.error('[WebSocket Client] Failed to parse message:', error);
      this.metrics.errors++;
    }
  }

  /**
   * Handle WebSocket close
   */
  private handleClose(code: number, reason: string): void {
    if (this.config.debug) {
      console.log(`[WebSocket Client] Connection closed: ${code} - ${reason}`);
    }

    this.clearTimers();
    this.ws = null;

    if (code !== 1000 && this.reconnectAttempts < this.config.maxReconnectAttempts) {
      this.scheduleReconnect();
    } else {
      this.setConnectionState(ConnectionState.DISCONNECTED);
    }

    this.emit('close', { code, reason });
  }

  /**
   * Handle WebSocket error
   */
  private handleError(error: Error): void {
    console.error('[WebSocket Client] WebSocket error:', error);
    this.metrics.errors++;
    this.setConnectionState(ConnectionState.ERROR);
    this.emit('error', error);
  }

  /**
   * Schedule reconnection attempt
   */
  private scheduleReconnect(): void {
    this.setConnectionState(ConnectionState.RECONNECTING);
    this.reconnectAttempts++;
    this.metrics.reconnections++;

    const delay = Math.min(
      this.config.reconnectInterval * Math.pow(2, this.reconnectAttempts - 1),
      30000 // Max 30 seconds
    );

    if (this.config.debug) {
      console.log(`[WebSocket Client] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts}/${this.config.maxReconnectAttempts})`);
    }

    this.reconnectTimer = setTimeout(() => {
      this.connect().catch(error => {
        console.error('[WebSocket Client] Reconnection failed:', error);
        if (this.reconnectAttempts < this.config.maxReconnectAttempts) {
          this.scheduleReconnect();
        } else {
          this.setConnectionState(ConnectionState.ERROR);
          this.emit('maxReconnectAttemptsReached');
        }
      });
    }, delay);
  }

  /**
   * Setup heartbeat mechanism
   */
  private setupHeartbeat(): void {
    this.heartbeatTimer = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.ping();
        
        // Check for stale connection
        const now = Date.now();
        if (this.lastHeartbeat > 0 && now - this.lastHeartbeat > this.config.heartbeatInterval * 2) {
          if (this.config.debug) {
            console.log('[WebSocket Client] Stale connection detected, reconnecting...');
          }
          this.ws.terminate();
        }
      }
    }, this.config.heartbeatInterval);
  }

  /**
   * Clear all timers
   */
  private clearTimers(): void {
    this.clearConnectionTimer();
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = undefined;
    }
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = undefined;
    }
  }

  /**
   * Clear connection timer
   */
  private clearConnectionTimer(): void {
    if (this.connectionTimer) {
      clearTimeout(this.connectionTimer);
      this.connectionTimer = undefined;
    }
  }

  /**
   * Set connection state and emit event
   */
  private setConnectionState(state: ConnectionState): void {
    if (this.connectionState !== state) {
      this.connectionState = state;
      this.emit('stateChange', state);
      
      if (this.config.debug) {
        console.log(`[WebSocket Client] State changed to: ${state}`);
      }
    }
  }

  /**
   * Get subscription key
   */
  private getSubscriptionKey(subscription: WSSubscription): string {
    const parts = [subscription.type];
    if (subscription.coin) parts.push(subscription.coin);
    if (subscription.user) parts.push(subscription.user);
    if (subscription.interval) parts.push(subscription.interval);
    return parts.join(':');
  }

  /**
   * Get default WebSocket endpoint for network
   */
  private getDefaultEndpoint(network: Network): string {
    return 'wss://api.hyperliquid.xyz/ws'; // Same endpoint for both mainnet and testnet
  }

  /**
   * Get connection state
   */
  getConnectionState(): ConnectionState {
    return this.connectionState;
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.connectionState === ConnectionState.CONNECTED;
  }

  /**
   * Get client metrics
   */
  getMetrics(): typeof this.metrics {
    return { ...this.metrics };
  }

  /**
   * Get active subscriptions
   */
  getSubscriptions(): string[] {
    return Array.from(this.subscriptions);
  }
}

File: sdks/python/hypersim_sdk/clients/websocket_client.py
"""
WebSocket streaming client with async context managers and automatic reconnection.
"""

import asyncio
import json
import logging
import time
from typing import Any, Dict, Optional, Callable, Set, AsyncGenerator, List
from contextlib import asynccontextmanager
import websockets
from websockets.exceptions import ConnectionClosed

from ..types.websocket import (
    WSConfig, WSMessage, WSSubscription, ConnectionState, WSMetrics,
    WSError, WSCloseInfo, SubscriptionType
)
from ..types.common import SubscriptionType as CommonSubscriptionType
from ..types.errors import NetworkError, TimeoutError, ValidationError, WebSocketError


logger = logging.getLogger(__name__)


class WebSocketClient:
    """Asynchronous WebSocket client with reconnection and subscription management."""
    
    def __init__(
        self,
        ws_url: str,
        config: Optional[WSConfig] = None,
        on_message: Optional[Callable[[WSMessage], None]] = None,
        on_connect: Optional[Callable[[], None]] = None,
        on_disconnect: Optional[Callable[[WSCloseInfo], None]] = None,
        on_error: Optional[Callable[[Exception], None]] = None
    ):
        """Initialize WebSocket client."""
        self.ws_url = ws_url
        self.config = config or WSConfig()
        
        # Event callbacks
        self.on_message = on_message
        self.on_connect = on_connect  
        self.on_disconnect = on_disconnect
        self.on_error = on_error
        
        # Connection state
        self._websocket: Optional[websockets.WebSocketServerProtocol] = None
        self._connection_state = ConnectionState.DISCONNECTED
        self._reconnect_count = 0
        self._running = False
        
        # Subscription management
        self._subscriptions: Dict[str, CommonSubscriptionType] = {}
        self._message_handlers: Dict[str, List[Callable[[Any], None]]] = {}
        
        # Tasks
        self._connection_task: Optional[asyncio.Task] = None
        
        # Metrics
        self._metrics = WSMetrics()
        self._start_time = time.time()
        
    @property
    def state(self) -> ConnectionState:
        """Get current connection state."""
        return self._connection_state
        
    @property
    def is_connected(self) -> bool:
        """Check if WebSocket is connected."""
        return (
            self._connection_state == ConnectionState.CONNECTED and
            self._websocket is not None and
            not self._websocket.closed
        )
        
    @property
    def metrics(self) -> WSMetrics:
        """Get connection metrics."""
        if self._connection_state == ConnectionState.CONNECTED:
            self._metrics.connection_uptime = time.time() - self._start_time
        return self._metrics
        
    async def connect(self) -> None:
        """Connect to WebSocket."""
        if self._running:
            logger.warning("WebSocket client is already running")
            return
            
        self._running = True
        self._connection_task = asyncio.create_task(self._connection_loop())
        
        # Wait for initial connection
        max_wait = self.config.connection_timeout
        start_time = time.time()
        
        while (
            self._connection_state != ConnectionState.CONNECTED and
            time.time() - start_time < max_wait
        ):
            await asyncio.sleep(0.1)
            
        if self._connection_state != ConnectionState.CONNECTED:
            raise TimeoutError(f"Failed to connect within {max_wait}s")
            
    async def disconnect(self) -> None:
        """Disconnect from WebSocket."""
        logger.info("Disconnecting WebSocket client")
        self._running = False
        
        if self._connection_task:
            self._connection_task.cancel()
            try:
                await self._connection_task
            except asyncio.CancelledError:
                pass
                
        if self._websocket:
            await self._websocket.close()
            
        self._connection_state = ConnectionState.DISCONNECTED
        self._websocket = None
        
    async def __aenter__(self) -> 'WebSocketClient':
        """Async context manager entry."""
        await self.connect()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.disconnect()
        
    async def _connection_loop(self) -> None:
        """Main connection loop with reconnection logic."""
        while self._running:
            try:
                self._connection_state = ConnectionState.CONNECTING
                logger.info(f"Connecting to {self.ws_url}")
                
                self._websocket = await asyncio.wait_for(
                    websockets.connect(
                        self.ws_url,
                        compression=None if not self.config.compression else "deflate",
                        max_size=self.config.max_message_size,
                    ),
                    timeout=self.config.connection_timeout
                )
                
                self._connection_state = ConnectionState.CONNECTED
                self._reconnect_count = 0
                
                logger.info("WebSocket connected successfully")
                
                if self.on_connect:
                    try:
                        self.on_connect()
                    except Exception as e:
                        logger.error(f"Error in connect callback: {e}")
                        
                # Resubscribe to all subscriptions
                await self._resubscribe_all()
                
                # Handle messages
                await self._handle_messages()
                
            except asyncio.TimeoutError:
                logger.error("WebSocket connection timeout")
                await self._handle_connection_error(TimeoutError("Connection timeout"))
                
            except ConnectionClosed as e:
                logger.warning(f"WebSocket connection closed: {e}")
                close_info = WSCloseInfo(code=e.code, reason=e.reason, was_clean=True)
                await self._handle_disconnect(close_info)
                
            except Exception as e:
                logger.error(f"WebSocket connection error: {e}")
                await self._handle_connection_error(e)
                
            # Reconnection logic
            if self._running and self.config.auto_reconnect:
                if self._reconnect_count < self.config.max_reconnect_attempts:
                    delay = min(
                        self.config.reconnect_interval * (2 ** self._reconnect_count),
                        60.0
                    )
                    logger.info(f"Reconnecting in {delay}s (attempt {self._reconnect_count + 1})")
                    
                    self._connection_state = ConnectionState.RECONNECTING
                    self._reconnect_count += 1
                    
                    await asyncio.sleep(delay)
                else:
                    logger.error("Max reconnection attempts reached")
                    self._running = False
                    
    async def _handle_messages(self) -> None:
        """Handle incoming WebSocket messages."""
        if not self._websocket:
            return
            
        try:
            async for message in self._websocket:
                try:
                    self._metrics.messages_received += 1
                    self._metrics.bytes_received += len(message)
                    self._metrics.last_activity = time.time()
                    
                    # Parse message
                    if isinstance(message, str):
                        data = json.loads(message)
                    else:
                        data = message
                        
                    # Create WSMessage
                    ws_message = WSMessage(
                        channel=data.get('channel', 'unknown'),
                        data=data.get('data', data),
                        message_type=data.get('type'),
                        sequence=data.get('sequence')
                    )
                    
                    if self.on_message:
                        try:
                            self.on_message(ws_message)
                        except Exception as e:
                            logger.error(f"Error in message callback: {e}")
                            
                    await self._dispatch_message(ws_message)
                    
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to parse message: {e}")
                    self._metrics.errors += 1
                except Exception as e:
                    logger.error(f"Error handling message: {e}")
                    self._metrics.errors += 1
                    
        except ConnectionClosed:
            logger.info("WebSocket message stream closed")
        except Exception as e:
            logger.error(f"Error in message handler: {e}")
            raise
            
    async def _dispatch_message(self, message: WSMessage) -> None:
        """Dispatch message to registered handlers."""
        handlers = self._message_handlers.get(message.channel, [])
        for handler in handlers:
            try:
                handler(message.data)
            except Exception as e:
                logger.error(f"Error in message handler for {message.channel}: {e}")
                
    async def _handle_connection_error(self, error: Exception) -> None:
        """Handle connection error."""
        self._connection_state = ConnectionState.ERROR
        self._metrics.errors += 1
        
        if self.on_error:
            try:
                self.on_error(error)
            except Exception as e:
                logger.error(f"Error in error callback: {e}")
                
    async def _handle_disconnect(self, close_info: WSCloseInfo) -> None:
        """Handle WebSocket disconnect."""
        self._connection_state = ConnectionState.DISCONNECTED
        
        if self.on_disconnect:
            try:
                self.on_disconnect(close_info)
            except Exception as e:
                logger.error(f"Error in disconnect callback: {e}")
                
    async def _resubscribe_all(self) -> None:
        """Resubscribe to all active subscriptions."""
        logger.info(f"Resubscribing to {len(self._subscriptions)} subscriptions")
        
        for sub_key, subscription in self._subscriptions.items():
            try:
                await self._send_subscription(subscription, subscribe=True)
            except Exception as e:
                logger.error(f"Failed to resubscribe to {sub_key}: {e}")
                
    async def _send_subscription(
        self,
        subscription: CommonSubscriptionType,
        subscribe: bool = True
    ) -> None:
        """Send subscription/unsubscription message."""
        if not self.is_connected:
            raise NetworkError("WebSocket not connected")
            
        message = {
            'method': 'subscribe' if subscribe else 'unsubscribe',
            'subscription': subscription.dict()
        }
        
        await self.send_message(message)
        
    async def send_message(self, message: Dict[str, Any]) -> None:
        """Send message to WebSocket."""
        if not self.is_connected or not self._websocket:
            raise NetworkError("WebSocket not connected")
            
        try:
            message_str = json.dumps(message)
            await self._websocket.send(message_str)
            
            self._metrics.messages_sent += 1
            self._metrics.bytes_sent += len(message_str)
            
        except ConnectionClosed:
            logger.error("WebSocket connection closed while sending message")
            raise NetworkError("Connection closed")
        except Exception as e:
            logger.error(f"Failed to send message: {e}")
            raise
            
    async def subscribe(
        self,
        subscription: CommonSubscriptionType,
        handler: Optional[Callable[[Any], None]] = None
    ) -> str:
        """Subscribe to data stream."""
        sub_key = self._generate_subscription_key(subscription)
        
        self._subscriptions[sub_key] = subscription
        
        if handler:
            if sub_key not in self._message_handlers:
                self._message_handlers[sub_key] = []
            self._message_handlers[sub_key].append(handler)
            
        if self.is_connected:
            await self._send_subscription(subscription, subscribe=True)
            
        self._metrics.active_subscriptions += 1
        logger.info(f"Subscribed to {sub_key}")
        
        return sub_key
        
    async def unsubscribe(self, subscription_key: str) -> None:
        """Unsubscribe from data stream."""
        if subscription_key not in self._subscriptions:
            logger.warning(f"Subscription {subscription_key} not found")
            return
            
        subscription = self._subscriptions[subscription_key]
        
        if self.is_connected:
            await self._send_subscription(subscription, subscribe=False)
            
        del self._subscriptions[subscription_key]
        if subscription_key in self._message_handlers:
            del self._message_handlers[subscription_key]
            
        self._metrics.active_subscriptions = max(0, self._metrics.active_subscriptions - 1)
        logger.info(f"Unsubscribed from {subscription_key}")
        
    def _generate_subscription_key(self, subscription: CommonSubscriptionType) -> str:
        """Generate unique subscription key."""
        parts = [subscription.type]
        
        if hasattr(subscription, 'coin') and subscription.coin:
            parts.append(subscription.coin)
        if hasattr(subscription, 'user') and subscription.user:
            parts.append(subscription.user)
        if hasattr(subscription, 'interval') and subscription.interval:
            parts.append(subscription.interval)
            
        return ':'.join(parts)
